File: test_client.py
import socket
import threading

import bs_rpc


class TestStreamMsg(bs_rpc.Body):
    def __init__(self, data):
        super().__init__("test.stream", data)


class TestRequestMsg(bs_rpc.Body):
    def __init__(self, data):
        super().__init__("test.request", data)


class TestResponseMsg(bs_rpc.Body):
    def __init__(self, data):
        super().__init__("test.response", data)


reg = bs_rpc.Registry()
reg.register(TestRequestMsg)
reg.register(TestResponseMsg)
reg.register(TestStreamMsg)

con = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
con.connect(("localhost", 9988))

client = bs_rpc.Client(con, reg)

listen = threading.Thread(
    target=lambda: client.listen({}),
    daemon=True)

listen.start()

print(client.call_for_response(
    TestRequestMsg({"Field1": "hmm", "Field2": "jaja"})).data)

print(client.call_for_response(
    TestRequestMsg({"Field1": "hmm", "Field2": "dock"})).data)

for resp in client.call_for_responses(TestStreamMsg({"Start": 7}))():
    print(resp.data)

File: pkg/bs_rpc/test/test_client.go
package test

import (
	"context"
	"log"
	"net"

	"github.com/lxgr-linux/pokete/bs_rpc"
	"github.com/lxgr-linux/pokete/bs_rpc/msg"
	msg1 "github.com/lxgr-linux/pokete/bs_rpc/test/msg"
)

func testClient() (results []msg.Body) {
	con, err := net.Dial("tcp", "localhost:9988")
	if err != nil {
		log.Panic(err)
	}

	reg, err := msg1.GetRegistry()
	if err != nil {
		log.Panic(err)
	}

	c := bs_rpc.NewClient(con, *reg)

	go func() {
		err = c.Listen(context.Background())
		if err != nil {
			log.Panic(err)
		}
	}()

	resp, err := c.CallForResponse(msg1.TestRequestMsg{"abc", "def"})
	if err != nil {
		log.Panic(err)
	}
	results = append(results, resp)

	ch, err := c.CallForResponses(msg1.TestStreamMsg{4})
	if err != nil {
		log.Panic(err)
	}
	for {
		val := <-ch
		if val == nil {
			break
		}
		results = append(results, val)
	}

	return
}
